pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -834,6 +840,91 @@ int kretprobe__inet6_bind(struct pt_regs* ctx) {
     return sys_exit_bind(ret);
 }
 
+SEC("kprobe/sockfd_lookup_light")
+int kprobe__sockfd_lookup_light(struct pt_regs* ctx) {
+    int sockfd = (int)PT_REGS_PARM1(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    bpf_map_update_elem(&sockfd_lookup_args, &pid_tgid, &sockfd, BPF_ANY);
+    return 0;
+}
+
+// this kretprobe is essentially creating an index of (PID, socketfd) to a conn_tuple_t object
+SEC("kretprobe/sockfd_lookup_light")
+int kretprobe__sockfd_lookup_light(struct pt_regs* ctx) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    int *sockfd = bpf_map_lookup_elem(&sockfd_lookup_args, &pid_tgid);
+    if (sockfd == NULL) {
+        return 0;
+    }
+
+    struct socket *socket = (struct socket*)PT_REGS_RC(ctx);
+    struct sock *skp;
+    bpf_probe_read(&skp, sizeof(skp), (char*)socket + offset_socket_sk());
+
+    conn_tuple_t t = {};
+    if (!read_conn_tuple(&t, skp, pid_tgid, CONN_TYPE_TCP)) {
+        goto cleanup;
+    }
+
+    u64 pid = pid_tgid >> 32;
+    u64 key = (pid << 32) | (*sockfd);
+    // TODO(before merging): Figure out how we want to cleanup this map;
+    // close(2) is an obvious candidate, but we would still need
+    // to have a TTL on map entries and expire them from userspace;
+    bpf_map_update_elem(&tup_by_pid_sockfd, &key, &t, BPF_ANY);
+
+cleanup:
+    bpf_map_delete_elem(&sockfd_lookup_args, &pid_tgid);
+    return 0;
+}
+
+SEC("kprobe/do_sendfile")
+int kprobe__do_sendfile(struct pt_regs* ctx) {
+    u64 fd_out = (int)PT_REGS_PARM1(ctx);
```

Comment:
Can we verify `out_fd` is a socket fd before updating the map?

Reply:
wouldn't that require a map lookup? what would be the benefit of that?

Reply:
We do the lookup either way (in the kprobe or kretprobe), but it would reduce the number of items we store in `do_sendfile_args` map at once. The "args" map could then store a `conn_tuple_t*` instead of the `fd` values.

It is also just a style choice of mine to filter early.

Reply:
This is done (as part of storing (struct sock*) as the value in the `sendfile_args_map`)